File: graph/15.cpp
#include <bits/stdc++.h>

using namespace std;

using Grid = vector<vector<int>>;
using Matrix = Grid;

struct sol
{
    pair<int, int> coordinates;
    int time;
};

bool isValid(Grid &G, Matrix &visited, int row, int col)
{
    return row >= 0 && row < G.size() && col >= 0 && col < G[0].size() && visited[row][col] == 0 && G[row][col] == 1;
}

int BFS(Grid &G, Matrix &visited, queue<sol> &q)
{
    vector<int> posRows = {-1, 0, 0, 1};
    vector<int> posCols = {0, -1, 1, 0};
    int time = 0;

    while (!q.empty())
    {
        auto v = q.front();
        q.pop();
        time = v.time;

        visited[v.coordinates.first][v.coordinates.second] = 1;

        for (int i = 0; i < 4; ++i)
        {
            int newPosRows = v.coordinates.first + posRows[i];
            int newPosCols = v.coordinates.second + posCols[i];

            if (isValid(G, visited, newPosRows, newPosCols))
            {
                G[newPosRows][newPosCols] = 2;
                q.push({{newPosRows, newPosCols}, time + 1});
            }
        }
    }

    return time;
}

bool checkMatrix(const Grid &G, bool checkFreshFruit, bool checkRottenFruit)
{
    for (int i = 0; i < G.size(); ++i)
    {
        for (int j = 0; j < G[i].size(); ++j)
        {
            if (checkFreshFruit && G[i][j] == 1)
                return true;
            if (checkRottenFruit && G[i][j] == 2)
                return true;
        }
    }

    return false;
}

int orangesRotting(vector<vector<int>> &grid)
{
    queue<sol> q;
    Matrix visited(grid.size(), vector<int>(grid[0].size(), 0));
    bool freshFruit = false;

    for (int i = 0; i < grid.size(); ++i)
    {
        for (int j = 0; j < grid[i].size(); ++j)
        {
            if (grid[i][j] == 2)
                q.push({{i, j}, 0});
        }
    }

    if (checkMatrix(grid, false, true) && checkMatrix(grid, true, false) == false)
        return 0;
    if (checkMatrix(grid, true, false) && checkMatrix(grid, false, true) == false)
        return -1;

    int answer = BFS(grid, visited, q);
    if (checkMatrix(grid, true, false))
        return -1;

    return answer;
}

int main()
{
    Grid G(1);
    G[0] = {0, 2};
    // G[1] = {0, 1, 1};
    // G[2] = {1, 0, 1};
    cout<<orangesRotting(G);

    return 0;
}

File: graph/15.py
from typing import List
from collections import deque


def is_valid(grid, visited, row, col):
    return (
        row >= 0
        and row < len(grid)
        and col >= 0
        and col < len(grid[0])
        and visited[row][col] == 0
        and grid[row][col] == 1
    )


def BFS(grid, visited, q: deque):
    time = 0

    posRows = [-1, 0, 0, 1]
    posCols = [0, -1, 1, 0]

    while len(q) != 0:
        row, col, time = q.popleft()
        visited[row][col] = 1

        for index in range(len(posRows)):
            newPosRow = row + posRows[index]
            newPosCol = col + posCols[index]

            if is_valid(grid, visited, newPosRow, newPosCol):
                grid[newPosRow][newPosCol] = 2
                q.append([newPosRow, newPosCol, time + 1])

    return time


def check_matrix(grid, check_fresh_fruit, check_rotten_fruit):
    for row in range(len(grid)):
        for col in range(len(grid[row])):
            if check_fresh_fruit is True and grid[row][col] == 1:
                return True
            if check_rotten_fruit is True and grid[row][col] == 2:
                return True

    return False


def orangesRotting(grid: List[List[int]]) -> int:
    q = deque()
    visited = [[0] * len(row) for row in grid]

    for row in range(len(grid)):
        for col in range(len(grid[row])):
            if grid[row][col] == 2:
                q.append([row, col, 0])

    if check_matrix(grid, True, False) and check_matrix(grid, False, True) is False:
        return -1
    if (
        check_matrix(grid, True, False) is False
        and check_matrix(grid, False, True) is True
    ):
        return 0

    answer = BFS(grid, visited, q)
    if check_matrix(grid, True, False):
        return -1

    return answer


grid = [[0, 2]]

print(orangesRotting(grid))
